modules/apps/callbacks/ibc_middleware.go

```diff
@@ -0,0 +1,243 @@
+		return appResult
+	}
+	if callbackData.SrcContractAddr == "" {
+		return appResult
+	}
+
+	cachedCtx, writeFn := ctx.CacheContext()
+	cachedCtx = cachedCtx.WithGasMeter(sdk.NewGasMeter(callbackData.GasLimit))
+
+	// call the contract
+	err = im.contractKeeper.IBCPacketTimeoutCallback(cachedCtx, packet, relayer, callbackData.SrcContractAddr)
+	if err == nil {
+		writeFn()
+	}
+	ctx.GasMeter().ConsumeGas(cachedCtx.GasMeter().GasConsumed(), "ibc packet timeout callback")
+
+	types.EmitSourceCallbackEvent(ctx, packet, types.CallbackTypeTimeoutPacket, callbackData, err)
+	return appResult
+}
+
+// OnChanCloseConfirm defers to the underlying application
+func (im IBCMiddleware) OnChanCloseConfirm(ctx sdk.Context, portID, channelID string) error {
+	return im.app.OnChanCloseConfirm(ctx, portID, channelID)
+}
+
+// OnChanCloseInit defers to the underlying application
+func (im IBCMiddleware) OnChanCloseInit(ctx sdk.Context, portID, channelID string) error {
+	return im.app.OnChanCloseInit(ctx, portID, channelID)
+}
+
+// OnChanOpenAck defers to the underlying application
+func (im IBCMiddleware) OnChanOpenAck(
+	ctx sdk.Context,
+	portID,
+	channelID,
+	counterpartyChannelID,
+	counterpartyVersion string,
+) error {
+	return im.app.OnChanOpenAck(ctx, portID, channelID, counterpartyChannelID, counterpartyVersion)
+}
+
+// OnChanOpenConfirm defers to the underlying application
+func (im IBCMiddleware) OnChanOpenConfirm(ctx sdk.Context, portID, channelID string) error {
+	return im.app.OnChanOpenConfirm(ctx, portID, channelID)
+}
+
+// OnChanOpenInit defers to the underlying application
+func (im IBCMiddleware) OnChanOpenInit(
+	ctx sdk.Context,
+	channelOrdering channeltypes.Order,
+	connectionHops []string,
+	portID,
+	channelID string,
+	channelCap *capabilitytypes.Capability,
+	counterparty channeltypes.Counterparty,
+	version string,
+) (string, error) {
+	return im.app.OnChanOpenInit(ctx, channelOrdering, connectionHops, portID, channelID, channelCap, counterparty, version)
+}
+
+// OnChanOpenTry defers to the underlying application
+func (im IBCMiddleware) OnChanOpenTry(
+	ctx sdk.Context,
+	channelOrdering channeltypes.Order,
+	connectionHops []string, portID,
+	channelID string,
+	channelCap *capabilitytypes.Capability,
+	counterparty channeltypes.Counterparty,
+	counterpartyVersion string,
+) (string, error) {
+	return im.app.OnChanOpenTry(ctx, channelOrdering, connectionHops, portID, channelID, channelCap, counterparty, counterpartyVersion)
+}
+
+// OnRecvPacket implements destination callbacks for the ibc-callbacks middleware.
+// It defers to the underlying application and then calls the contract callback.
+// If the contract callback fails (within the gas limit), state changes are reverted.
+func (im IBCMiddleware) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, relayer sdk.AccAddress) ibcexported.Acknowledgement {
+	appAck := im.app.OnRecvPacket(ctx, packet, relayer)
+
+	appAckResult, ok := appAck.(channeltypes.Acknowledgement)
```

Comment:
Again you are casting to the default type which will fail if the underlying app is using custom acks